### PR TITLE
[Terrain] First pass at an image gradient painter

### PR DESCRIPTION
**Pull request by mbalfour-amzn**
## What does this PR do?

NOTE: This code was originally written by @its-tn10 .

This is the first pass at adding a "painting" editing mode to the Image Gradient. The feature is still gated by a .setreg setting and is hidden by default.

When the "Edit" button is pressed, a paint brush circle appears in the viewport that can be used to paint values into the current image gradient. The paint brush has controls for radius, intensity, and opacity. These controls can currently be edited on the component UX underneath the Edit button. 

Implementation notes:
* EditorBaseComponentMode needed to inherit from ToolsApplicationNotificationBus as protected instead of private due to a quirk of EBus implementation. If a subclass of EditorBaseComponentMode wants to _send_ a notification on ToolsApplicationNotificationBus (in particular, to refresh the component controls), the private inheritance of that bus causes a compiler error, even if the base class hasn't implemented that particular notification. Switching it to protected fixes the issue.
* **ManipulatorViewProjectedCircle** is a new view for drawing projected circles. Any manipulator can use it.
* **PaintBrushManipulator** is a generic implementation of a paint brush manipulator. It provides all of the logic for painting in the viewport, but leaves the actual data manipulation from the painting up to the component mode that's using the manipulator.
* **EditorImageGradientComponentMode** implements a specific component mode that uses the paint brush to paint values into an image gradient.

Known issues:
* The paint brush settings are overly complex right now due to being displayed on the component card. The goal is to eventually move these controls in-viewport, which will simplify the handling of the settings and provide a better UX experience.
* Undo/redo might not work correctly.
* Streaming image assets don't currently hot reload correctly, so after editing is completed, it's possible that the image will temporarily revert to a previous state. Reloading the level works around the issue.
* The UX flow for saving hasn't been investigated yet.

## How was this PR tested?

Tested out the paint brush tools manually in a level where image gradients control terrain height, terrain surface types, and vegetation placement.

https://user-images.githubusercontent.com/82224783/186214056-b150779f-ea37-4b7f-bec8-f1080adee115.mp4

### CODE REVIEW

**Review by cgalvan — APPROVED** [review]

File: Code/Framework/AzToolsFramework/AzToolsFramework/Manipulators/PaintBrushManipulator.cpp [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+    }
+
+    AZ::u32 PaintBrushConfig::OnRadiusChange()
+    {
+        // Notify listeners that the configuration changed. This is used to synchronize the internal manipulator configuration
+        // with the visible editable one stored on the Editor component.
+        PaintBrushNotificationBus::Event(m_ownerEntityComponentId, &PaintBrushNotificationBus::Events::OnRadiusChanged, m_radius);
+        return AZ::Edit::PropertyRefreshLevels::ValuesOnly;
+    }
+
```
Comment (jonawals) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Manipulators/PaintBrushManipulator.cpp:81]: Nit: could remove some of these extra line breaks?

File: Code/Framework/AzToolsFramework/AzToolsFramework/Manipulators/ManipulatorView.h [hunk 2]
```diff
@@ -304,6 +305,30 @@ namespace AzToolsFramework
...
+            AzFramework::DebugDisplayRequests& debugDisplay,
+            const AzFramework::CameraState& cameraState,
+            const ViewportInteraction::MouseInteraction& mouseInteraction) override;
+
+        void SetRadius(float radius);
+
+        AZ::Vector3 m_axis = AZ::Vector3::CreateAxisZ();
+        float m_width = 0.5f;
+        float m_radius = 2.0f;
+        AZ::Color m_color = AZ::Color(1.0f, 0.0f, 0.0f, 1.0f);
```
Comment (jonawals) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Manipulators/ManipulatorView.h:329]: Suggestion: could we use `AZ::Colors::Red` here instead?

File: Code/Framework/AzToolsFramework/AzToolsFramework/Manipulators/PaintBrushManipulator.cpp [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+        const AZ::Transform& worldFromLocal, const AZ::EntityComponentIdPair& entityComponentIdPair)
+    {
+        m_ownerEntityComponentId = entityComponentIdPair;
+        m_config.m_ownerEntityComponentId = entityComponentIdPair;
+        PaintBrushRequestBus::Handler::BusConnect(entityComponentIdPair);
+
+        SetSpace(worldFromLocal);
+
+        // The PaintBrush manipulator uses a circle projected into world space to represent the brush.
+        const AZ::Color manipulatorColor = AZ::Color(1.0f, 0.0f, 0.0f, 1.0f);
```
Comment (jonawals) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Manipulators/PaintBrushManipulator.cpp:100]: Suggestion: could we use `AZ::Colors::Red` here instead?

File: Gems/GradientSignal/Code/Source/Components/ImageGradientComponent.cpp [hunk 4]
```diff
@@ -997,6 +1000,24 @@ namespace GradientSignal
...
+
+        if (width > 0 && height > 0)
+        {
+            const AZ::Aabb bounds = m_gradientTransform.GetBounds();
+            const AZ::Vector2 boundsMeters(bounds.GetExtents());
+            const AZ::Vector2 imagePixelsInBounds(width / GetTilingX(), height / GetTilingY());
+            return imagePixelsInBounds / boundsMeters;
+        }
+
+        return AZ::Vector2(0.0f, 0.0f);
```
Comment (jonawals) [inline Gems/GradientSignal/Code/Source/Components/ImageGradientComponent.cpp:1018]: Suggestion: could we use `AZ::Vector2::CreateZero()` instead? I know it's a bit more verbose but potentially more idiomatic?

**Review by jonawals — APPROVED** [review]

**Pull request merged**